File: CPP/No802.cc
/**
 * Created by Xiaozhong on 2020/12/23.
 * Copyright (c) 2020/12/23 Xiaozhong. All rights reserved.
 */
#include <vector>
#include <iostream>

using namespace std;

class Solution {
public:
    vector<int> eventualSafeNodes(vector<vector<int>> &graph) {
        int n = graph.size();
        vector<int> color(n);
        vector<int> ans;
        for (int i = 0; i < n; ++i) if (dfs(i, color, graph)) {
            ans.push_back(i);
        }
        return ans;

    }

private:
    // black: 2, gray: 1, white: 0
    bool dfs(int node, vector<int> &color, vector<vector<int>> &graph) {
        if (color[node] > 0) return color[node] == 2;
        color[node] = 1;
        for (int neighboor : graph[node]) {
            if (color[node] == 2) continue;
            if (color[neighboor] == 1 || !dfs(neighboor, color, graph)) return false;
        }
        color[node] = 2;
        return true;
    }
};

File: Java/src/No802.java
/*
 * Copyright (c) 8/5/2021, HuXiaozhong. All rights reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 */

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class No802 {
    private enum Color {
        WHITE, GRAY, BLACK
    }

    public List<Integer> eventualSafeNodes(int[][] graph) {
        List<Integer> ans = new ArrayList<>();
        Color[] colors = new Color[graph.length];
        Arrays.fill(colors, Color.WHITE);
        for (int start = 0; start < graph.length; ++start) {
            if (dfs(graph, colors, start)) ans.add(start);
        }
        return ans;
    }

    private boolean dfs(int[][] graph, Color[] colors, int start) {
        if (colors[start] != Color.WHITE) return colors[start] == Color.BLACK;
        colors[start] = Color.GRAY;
        for (int next : graph[start]) {
            if (!dfs(graph, colors, next)) return false;
        }
        colors[start] = Color.BLACK;
        return true;
    }
}
